builder/builder.go

```diff
@@ -0,0 +1,390 @@
+	if b.err != nil {
+		return b
+	}
+
+	s = escapeSpecialChars(s)
+
+	if len(b.variants) == 0 {
+		b.patterns = append(b.patterns, s)
+	} else {
+		b.variants[len(b.variants)-1].pattern = append(b.variants[len(b.variants)-1].pattern, s)
+	}
+
+	return b
+}
+
+func (b *Builder) Local(v string, expr *Expression) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	if len(b.patterns) > 0 {
+		b.err = fmt.Errorf("complex message cannot be added after simple message")
+		return b
+	}
+
+	expr.spacing = b.spacing
+
+	b.locals = append(b.locals, local{variable: variable(v), expr: expr})
+
+	return b
+}
+
+func (b *Builder) Input(expr *Expression) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	if len(b.patterns) > 0 {
+		b.err = fmt.Errorf("complex message cannot be added after simple message")
+		return b
+	}
+
+	expr.spacing = b.spacing
+
+	b.inputs = append(b.inputs, *expr)
+
+	return b
+}
+
+func (b *Builder) Expr(e *Expression) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	e.spacing = b.spacing
+
+	if len(b.variants) == 0 {
+		b.patterns = append(b.patterns, e)
+	} else {
+		b.variants[len(b.variants)-1].pattern = append(b.variants[len(b.variants)-1].pattern, e)
+	}
+
+	return b
+}
+
+func (b *Builder) Match(selectors ...*Expression) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	if len(b.patterns) > 0 {
+		b.err = fmt.Errorf("complex message cannot be added after simple message")
+		return b
+	}
+
+	if len(selectors) == 0 {
+		b.err = fmt.Errorf("match MUST have at least one selector")
+		return b
+	}
+
+	for i := range selectors {
+		b.selectors = append(b.selectors, *selectors[i])
+	}
+
+	return b
+}
+
+func (b *Builder) Key(keys ...any) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	if len(keys) != len(b.selectors) {
+		b.err = fmt.Errorf("number of keys in each variant MUST match the number of selectors in the matcher")
+		return b
+	}
+
+	b.variants = append(b.variants, variant{keys: keys})
+
+	return b
+}
+
+type variant struct {
+	keys, pattern []any
+}
+
+func (v *variant) String() string {
+	var s string
+
+	for i := range v.keys {
+		switch v := v.keys[i].(type) {
+		case string:
+			s += v + " "
+		case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
+			s += fmt.Sprintf("%d ", v)
+		case float32, float64:
+			s += fmt.Sprintf("%f ", v)
+		default:
+			panic(fmt.Sprintf("unsupported key type: %T", v))
+		}
+	}
+
+	s += "{{"
+
+	for i := range v.pattern {
+		if i != 0 {
+			s += " "
+		}
+
+		switch v := v.pattern[i].(type) {
+		case string:
+			s += v
+		case *Expression:
+			s += v.String()
+		default:
+			panic(fmt.Sprintf("unsupported pattern type: %T", v))
+		}
+	}
+
+	return s + "}}"
+}
+
+type local struct {
+	expr     *Expression
+	variable variable
+}
+
+type literal any
+
+type variable string
+
+type function struct {
+	name    string
+	options []option
+}
+
+type Expression struct {
+	spacing   string
+	operand   any // literal or variable
+	functions []function
```

Comment:
Expression can have only one function
```abnf
annotation = (function *(s option))
```

\+ and \- functions, are two separate expressions
hello {+func1} world {-func2}